src/cli.c

```diff
@@ -1502,6 +1571,439 @@ process_result(SSL* ssl, int socket, int32_t output_format)
+{
+   struct json* read = NULL;
+
+   if (pgmoneta_management_read_json(ssl, socket, NULL, NULL, &read))
+   {
+      goto error;
+   }
+
+   if (get_config_key_result(config_key, read))
+   {
+      goto error;
+   }
+
+   if (MANAGEMENT_OUTPUT_FORMAT_RAW != output_format)
+   {
+      translate_json_object(read);
+   }
+
+   if (MANAGEMENT_OUTPUT_FORMAT_TEXT == output_format)
+   {
+      pgmoneta_json_print(read, FORMAT_TEXT);
+   }
+   else
+   {
+      pgmoneta_json_print(read, FORMAT_JSON);
+   }
+
+   pgmoneta_json_destroy(read);
+
+   return 0;
+
+error:
+
+   pgmoneta_json_destroy(read);
+
+   return 1;
+}
+
+static int
+get_config_key_result(char* config_key, struct json* j)
+{
+   char context[MISC_LENGTH];
+   char server[MISC_LENGTH];
+   char key[MISC_LENGTH];
+
+   struct json* configuration_js;
+   uintptr_t config_value;
+   struct json* response = NULL;
+   bool main_section = false;
+   bool server_section = false;
+   int begin = -1, end = -1;
+
+   memset(context, 0, MISC_LENGTH);
+   memset(server, 0, MISC_LENGTH);
+   memset(key, 0, MISC_LENGTH);
+
+   for (int i = 0; i < strlen(config_key); i++)
+   {
+      if (config_key[i] == '.')
+      {
+         if (!strlen(context))
+         {
+            memcpy(context, &config_key[begin], end - begin + 1);
+            context[end - begin + 1] = '\0';
+            begin = end = -1;
+            continue;
+         }
+         else if (!strlen(server))
+         {
+            memcpy(server, &config_key[begin], end - begin + 1);
+            server[end - begin + 1] = '\0';
+            begin = end = -1;
+            continue;
+         }
+      }
+
+      if (begin < 0)
+      {
+         begin = i;
+      }
+
+      end = i;
+
+   }
+
+   // if the key has not been found, since there is no ending dot,
+   // try to extract it from the string
+   if (!strlen(key))
+   {
+      memcpy(key, &config_key[begin], end - begin + 1);
+      key[end - begin + 1] = '\0';
+   }
+
+   // force the main section, i.e., global parameters, if and only if
+   // there is no section or section is 'pgmoneta' without any subsection
+   server_section = (strlen(context) && !strncmp(context, MANAGEMENT_ARGUMENT_SERVERS_CONFIGURATION, MISC_LENGTH) && strlen(server));
+   main_section = (strlen(context) && !strncmp(context, MANAGEMENT_ARGUMENT_MAIN_CONFIGURATION, MISC_LENGTH));
+
+   response = (struct json*)pgmoneta_json_get(j, MANAGEMENT_CATEGORY_RESPONSE);
+
+   if (!response || !(main_section || server_section))
+   {
+      goto error;
+   }
+
+   if (!main_section)
+   {
+      configuration_js = (struct json*)pgmoneta_json_get(response, context);
+      if (!configuration_js)
+      {
+         goto error;
+      }
+      configuration_js = (struct json*)pgmoneta_json_get(configuration_js, server);
+   }
+   else
+   {
+      configuration_js = (struct json*)pgmoneta_json_get(response, context);
+   }
+
+   if (!configuration_js)
+   {
+      goto error;
+   }
+
+   if (!strcmp(key, "host"))
```

Comment:
This looks like a PITA.

An idea is that you can use iterator. Our json iterator works not only for json arrays but for json items as well. So just create an iterator and traverse the `configuration_js`, and put the string of the value into the response directly.

So something like:
```
if (!strcmp(key, iter->key) {
   pgmoneta_json_put(response, key, (uintptr_t)pgmoneta_value_to_string(iter->value), ValueString);
}
```
It'll save a lot of trouble for you. 

Sorry for proposing this 2 days later, after you've been through this... I've been busy with work recently.

Feel free to take the idea and see if you can make other parts simpler.

Reply:
@Jubilee101, This was really helpful. Thanks a lot!